stm/asyncTaskQueue.mli

```diff
@@ -6,79 +6,129 @@
+
+      The data must be marshallable as it send through the network
+      using [Mashal] . *)
   type request
   type response
 
-  val name : string ref (* UID of the task kind, for -toploop *)
+  (** UID of the task kind, for -toploop *)
+  val name : string ref
+  (** Extra arguments of the task kind, for -toploop *)
   val extra_env : unit -> string array
 
-  (* run by the master, on a thread *)
+  (** {5} Client API, it is run by the master, on a thread *)
+
+  (** [request_of_task status tid] checks if the worker has already
+      processed [tid] *)
   val request_of_task : competence worker_status -> task -> request option
+
+  (** [task_match status tid] XXX *)
   val task_match : competence worker_status -> task -> bool
+
+  (** [use_response status tid out] notifies the worker that [tid] can
+      be freeded? *)
   val use_response :
     competence worker_status -> task -> response ->
       [ `Stay of competence * task list | `End ]
+
+  (** [on_marshal_error err_msg tid] notifies of marshaling failure. *)
   val on_marshal_error : string -> task -> unit
+
+  (** [on_task_cancellation_or_expiration_or_slave_death tid] XXX  *)
   val on_task_cancellation_or_expiration_or_slave_death : task option -> unit
+
+  (** [forward_feedback fb] sends fb to all the workers. *)
   val forward_feedback : Feedback.feedback -> unit
- 
-  (* run by the worker *)
+
+  (** {5} Server API, it is run by worker, on a different fresh process *)
+
+  (** [perform in] synchronously processes  *)
   val perform : request -> response
 
-  (* debugging *)
+  (** debugging *)
   val name_of_task : task -> string
   val name_of_request : request -> string
 
 end
 
+(** XXX *)
 type expiration = bool ref
 
+(** Client-side functor. [MakeQueue T] creates a task queue for task [T]  *)
 module MakeQueue(T : Task) : sig
 
+  (** [queue] is the abstract queue type. *)
   type queue
 
-  (* Number of workers, 0 = lazy local *)
+  (** [create n] will initialize the queue with [n] workers. If [n] is
+      0, the queue won't spawn any process, working in a lazy local
+      manner. *)
   val create : int -> queue
+
+  (** [destroy q] Deallocates [q], cancelling all pending tasks. *)
   val destroy : queue -> unit
 
+  (** [n_workers q] returns the number of workers of [q] *)
   val n_workers : queue -> int
 
+  (** [enqueue_task q t p] schedules [t] for execution in [q]. [p] does XXX *)
   val enqueue_task : queue -> T.task * expiration -> unit
 
-  (* blocking function that waits for the task queue to be empty *)
+  (** [join q] blocks until the task queue is empty *)
   val join : queue -> unit
+
+  (** [cancel_all q] Cancels all tasks *)
   val cancel_all : queue -> unit
 
+  (** [cancel_worker q wid] cancels a particular worker [wid] *)
   val cancel_worker : queue -> WorkerPool.worker_id -> unit
 
+  (** [set_order q cmp] reorders [q] using ordering [cmp] *)
   val set_order : queue -> (T.task -> T.task -> int) -> unit
 
+  (** [broadcast q] XXX *)
```

Comment:
This is nasty. As I was saying workers can be picky, e.g. pick tasks only when they are "on screen".
Of course the screen is scrolled, and that changes the potential choice of workers to pick up a task or not.

This function wakes up the workers (the managers) that give a look (again) to the tasks in the queue.

The STM calls it when the perspective (as in PIDE) changes.

Reply:
What sucks here is that why `task_match` has access to the competence data in order to decide if the task is palatable to the worker or not... such data is local to the worker (manager).
The perspective is global, so it does not quite fit this picture.  This API to make all managers reconsider the tasks in the queue is the best I could came up with.

This API is crucial to Coqoon (or any other UI that invokes Stm.finish eagerly but wants the workers to "focus" on the visible part of the document).